### DISCUSSION TITLE
Is it possible to disable build_ext for develop/editable installation?

### DISCUSSION THREAD

**Discussion by senysenyseny16** [pypa/setuptools#3911] [Q&A] [answered]
hi!

Most of the python source files (*.py) are compiled into *.so for "production" wheel, but for development it is necessary to keep them as source files.

I tried to implement `SubCommand` protocol for `build_ext` command with `self.editable_mode` in `run, initialize/finalize_options` methods, something like:
```python3
if self.editable_mode:
    return
```
But when I type `pip install -vvv -e .`, `initialize/finalize options` methods of `build_ext` command are triggered many times and only in the last call `self.editable_mode` is True:
```
1. Running command Getting requirements to build editable | (cython here)
2. Running command Preparing editable metadata (pyproject.toml) | (cython here)
3. Running command Building editable for <package-name> (pyproject.toml) | (cython x2)
```
So, I have a lot of extra work and unwanted artifacts in the source directory.

How can i solve this problem?
Thanks in advance.

**Comment by abravalheri** [comment]
Have you considered postponing some of the processing you are doing to the `run` method?

In theory the `initialize/finalize_options` methods should not generate artifacts or modify the file system, only the `run` file is assumed to do that.

The `if self.editable_mode` trick should work for `run`, but there is no guarantees in `initialize/finalize_options`.

**Reply by senysenyseny16** [reply]
Hi @abravalheri, thanks for your reply.

In simplified case I need two behavior:

- "production": all `*.py` files should be compiled into `*.so` with the help of Cython (`*.py -> *.c -> *.so`)
- editable: nothing to compile, no `build_ext/ext_modules`

To make "production" wheel I know two ways, the first of them:
```python3
setup(
    ext_modules = cythonize("package/*.py")
)
```
In this case Cython will compile extensions for any installation (prod and editable).

The second: extend `build_ext` command from Cython package, but it intensively uses `initialize/finalize options`.

**Reply by abravalheri** [reply, marked as answer]
If you have a look on the source-code, you can see that both [`setuptools.command.build_ext:build_ext`](https://github.com/pypa/setuptools/blob/main/setuptools/command/build_ext.py) and [`Cython.Distutils.build_ext:build_ext`](https://github.com/cython/cython/blob/master/Cython/Distutils/build_ext.py) don't create any files in `initialize/finalize_options`, they just modify attributes.

They will only touch the disk when the `run` method is called. With that in mind, it would still be possible to customise the `run` method behaviour based on `if self.editable_mode`...

So maybe the extra artifacts that you don't want are generated elsewhere?

How about the `cythonize` function you are calling in `setup.py`. Are you sure is not this function that changes the file system?

---

> In this case Cython will compile extensions for any installation (prod and editable).

Have you tried extending `build_ext` and using `if self.editable_mode` inside `def run` to avoid the compilation, what is the outcome?

Please remember to set `self.editable_mode = False` inside `initialize_options` so the attribute is defined. This will allow setuptools to automatically modify it later.

**Reply by senysenyseny16** [reply]
Thank you @abravalheri. Problem solved.

**Reply by Vegal56** [reply]
Thanks for solution
!!!
